src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -55,7 +55,7 @@ public Memory<T> Slice(int index, int length)
 
         public Span<T> Span => _owner.GetSpanInternal(_id).Slice(_index, _length);
 
-        public DisposableReservation Reserve() => new DisposableReservation(_owner, _id);
+        public DisposableReservation<T> Reserve() => new DisposableReservation<T>(_owner, _id);
```

Comment:
So if DisposableReservation is non-generic then it can't give you access to the Span directly.  Could you elaborate a little on the use case, as storing reservations in a collection to me sounds dangerous, particularly with the ReferenceCounter.